dmn-engine/src/main/scala/org/camunda/dmn/evaluation/DecisionTableEvaluator.scala

```diff
@@ -45,42 +46,67 @@ class DecisionTableEvaluator(
 
   def eval(decisionTable: ParsedDecisionTable,
            context: EvalContext): Either[Failure, Val] = {
-    implicit val ctx = context
-
-    evalInputExpressions(decisionTable.inputs).right.flatMap { inputValues =>
-      checkRules(decisionTable.rules, inputValues).right
-        .map(_.flatten)
-        .flatMap { matchedRules =>
-          matchedRules match {
-            case Nil => {
-              applyDefaultOutputEntries(decisionTable.outputs).right.map {
-                result =>
+    implicit val ctx: EvalContext = context
+
+    evalInputExpressions(decisionTable.inputs) match {
+      case Left(failure) =>
+        audit(decisionTable, Nil, Nil, ValError(failure.message))
+        Left(failure)
+      case Right(inputValues) =>
+        (checkRules(decisionTable.rules, inputValues) match {
+          case Left(failure) =>
+            audit(decisionTable, inputValues, Nil, ValError(failure.message))
+            Left(failure)
+          case right =>
+            right
+        }).map(_.flatten)
+          .flatMap {
+            case Nil =>
+              applyDefaultOutputEntries(decisionTable.outputs) match {
```

Comment:
Unfortunately, this change makes it really hard to see the actual business logic. We should find a way to keep the readability. 

Maybe, we can extract the failure handling, or transform it in an easier structure.

Reply:
With adding the same pattern as I did in audit it looks now like this https://github.com/camunda/dmn-scala/pull/16/commits/bbd3ddc13c73fd0bdcc25d9ad948b0e4e595511b

Still space for improvement, but I think now it is at least readable.

Reply:
It is better :sweat_smile: But I think it is still hard to read with all the pattern matching in between. I would like to keep the whole audit out of the core business logic :thinking: I'll try to make a suggestion.